src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -0,0 +1,365 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.PatternMatching;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.LanguageServices.Implementation.LanguageServiceBrokerShim;
+using Microsoft.VisualStudio.Text;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.VisualStudio.LanguageServices
+{
+    internal static class DocumentOutlineHelper
+    {
+        /// <summary>
+        /// The text snapshot from when the document symbol request was made.
+        /// </summary>
+        private static ITextSnapshot? LspSnapshot { get; set; }
+
+        /// <summary>
+        /// The text snapshot from the last active text view.
+        /// </summary>
+        private static ITextSnapshot? CurrentSnapshot { get; set; }
+
+        /// <summary>
+        /// Is true when LspSnapshot and CurrentSnapshot have been initialized.
+        /// </summary>
+        [MemberNotNullWhen(true, nameof(LspSnapshot), nameof(CurrentSnapshot))]
+        private static bool IsSnapshotInitialized { get; set; }
+
+        /// <summary>
+        /// The caret position of the last active text view.
+        /// </summary>
+        private static int CaretPosition { get; set; }
+
+        /// <summary>
+        /// Given an array of all Document Symbols in a document, returns an array containing the 
+        /// top-level Document Symbols and their nested children.
+        /// </summary>
+        public static DocumentSymbol[] GetNestedDocumentSymbols(DocumentSymbol[]? documentSymbols)
+        {
+            if (documentSymbols is null || documentSymbols.Length == 0)
+                return Array.Empty<DocumentSymbol>();
+
+            var allSymbols = documentSymbols
+                .SelectMany(x => x.Children)
+                .Concat(documentSymbols)
+                .OrderBy(x => x.Range.Start.Line)
+                .ThenBy(x => x.Range.Start.Character)
+                .ToImmutableArray();
+
+            return GroupDocumentSymbolTrees(allSymbols)
+                .Select(group => CreateDocumentSymbolTree(group))
+                .ToArray();
```

Comment:
So this looks like we're taking a list of symbols that are nested (given there's a "Children" collection), we're flattening it out, and then regrouping them back into a hierarchy again?

Reply:
@jasonmalinowski the LSP request unfortunately only returns 2 levels (a node and its children) and doesn't do any further nesting. That's why this code is necessary. Please let me know if you think it's a good idea to add docs that explain this until this is fixed.

Reply:
Ah yes, a comment would be great here!

Reply:
And is that something being looked at generally to be fixed in LSP?

Reply:
I'll be glad to add docs explaining that here! 

And as for if it's something looked at to be fixed in LSP, I'm sure @dibarbet will have more context here :)